### ISSUE TITLE
Nesting @Transaction with @CreateSqlObject doesn't work so well

### ISSUE THREAD

**Issue by stevenschlansker** [jdbi/jdbi#855]
Via email:

> I have the following code whose transaction works, i.e. if the insert into the roles table fails, then there is no user created in the user table.

```java
UserService service = dbi.onDemand(UserService.class);

public abstract class UserService {
  @CreateSqlObject
  abstract UserDao userDao();

  public Long createUser(final String username, final List<String> roles) {
    return userDao().createUser(username, roles);
  }
}

public abstract class UserDao implements Closeable {

  @SqlUpdate("insert into `user` (username) values (:username)")
  @GetGeneratedKeys
  public abstract Long create(@Bind("username") String username);

  @SqlBatch("insert into `role` (role) values (:role)")
  public abstract void insertRoles(@Bind("user_id") String userId, @Bind("role") List<String> roles);

  @Transaction
  public Long createUser(final String username, final List<String> roles) {
    Long userId = create(username);
    insertRoles(userId, roles);
    return userId;
  }
}
```

> However, if I modify the code to the following, (which is what I want), the transaction doesn't work. I.e. if the insertion into the roles table fails, the user inserted into the user table is still there. Why doesn't the code below work, and what would I need to do to make it work? The reason I want to do it the way below is if I decide to create another DAO class, then I would need the transaction to work across methods in different DAOs.

```java
UserService service = dbi.onDemand(UserService.class);

public abstract class UserService {
  @CreateSqlObject
  abstract UserDao userDao();

  @Transaction
  public Long createUser(final String username, final List<String> roles) {
    UserDao userDao = userDao();
    Long userId = userDao.create(username);
    userDao.insertRoles(userId, roles);
    return userId;
  }
}

public abstract class UserDao implements Closeable {

  @SqlUpdate("insert into `user` (username) values (:username)")
  @GetGeneratedKeys
  public abstract Long create(@Bind("username") String username);

  @SqlBatch("insert into `role` (role) values (:role)")
  public abstract void insertRoles(@Bind("user_id") String userId, @Bind("role") List<String> roles);
}
```

**Comment by jarlah** [comment]
I came here because I have the exact same problem. Or similar.

I have a service:

```
public class LocationServicce {
  final LocationDao locationDao;
  LocationService(Jdbi jdbi) {
    this.locationDao = jdbi.onDemand(LocationDao.class);
  }
}
```

I have a dao:

```
public interface LocationDao {
  @CreateSqlObject
  SomeOtherDao otherDao;

  void createStuff(...):

  @Transaction
  default void doStuffInTransaction(...) {
     createStuff(..)
     otherDao.createOtherStuff();
  }
}
```

If createStuff succeeds and createOtherStuff succeeds, the result of createStuff is not rolled back.

Is this a problem? Am I doing it wrong?

And, if I its a problem, is it difficult to fix?

**Comment by qualidafial** [comment]
This slipped past my radar.

I'm actually a little mystified that this _doesn't_ work as expected.

When you call an on-demand method, it is like calling `useExtension(MyDao.class, dao -> dao.doStuffInTransaction())`. In theory, calling `@CreateSqlObject` from that should share the same handle.

Um.. what happens if you call `getHandle()` first in the transactional method--and just do nothing with it? I have a hunch this might have to do with `LazyHandleSupplier` type that we use to defer allocating a handle until a method is called that actually needs it.

**Comment by TheRealMarnes** [comment]
I've pushed a test case for this, but the test is green, meaning transactions across `@Transaction` and `@CreateSqlObject` seem to work. Can anyone confirm if this issue is still valid by making the test case fail?

**Comment by TheRealMarnes** [comment]
... or is this a v2 issue? I see abstract classes instead of interfaces being used in the example, and there's a v2 milestone set...

Didn't know v2 code could look so identical to v3 code. 😄

**Comment by stevenschlansker** [comment]
Hi @jarlah can you take a look at the test case Marnes posted above and see if your issue is fixed in Jdbi 3?  If not, can you clarify how your use differs from the test case posted?  Thanks!

**Comment by stevenschlansker** [comment]
This should be fixed in #1484 once and for all.  The v2 behavior is not going to change - if this is a problem for you, please upgrade, and if the new behavior doesn't work for you let's open a new ticket to fix it :)

**Comment by TheRealMarnes** [comment]
Posting the test code here and deleting the branch so it's out of our way in the repo:

```
package org.jdbi.v3.sqlobject;

import java.util.List;
import org.jdbi.v3.sqlobject.statement.SqlQuery;
import org.jdbi.v3.sqlobject.statement.SqlUpdate;
import org.jdbi.v3.sqlobject.transaction.Transaction;
import org.jdbi.v3.testing.JdbiRule;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatCode;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class NestedTransactionTest {
    @Rule
    public JdbiRule db = JdbiRule.sqlite().withPlugin(new SqlObjectPlugin());

    @Before
    public void before() {
        db.getHandle().createUpdate("create table foo(bar int primary key)").execute();
    }

    @Test
    public void nestedTransactionsWork() {
        Top top = db.getHandle().attach(Top.class);

        assertThat(top.select()).isEmpty();

        assertThatThrownBy(top::breakTx).isNotNull();

        assertThat(top.select()).isEmpty();
    }

    public interface Top {
        @CreateSqlObject
        Sub sub();

        @SqlUpdate("insert into foo(bar) values(1)")
        void insert();

        @SqlQuery("select bar from foo")
        List<Integer> select();

        @Transaction
        default void breakTx() {
            Top top = this;
            Sub sub = sub();

            assertThatCode(top::insert).doesNotThrowAnyException();
            assertThat(select()).containsExactly(1);

            sub.insert();
            Assert.fail("should have thrown");
        }
    }

    public interface Sub {
        @SqlUpdate("insert into foo(bar) values(1)")
        void insert();
    }
}
```
